### ISSUE TITLE
Syntax Highlighting in Escaped String Literals

### ISSUE THREAD

**Issue by jpe90** [tree-sitter/tree-sitter-haskell#42]
Hello,

Is it possible to highlight escaped string literals? Here's how they look currently:

![126541203-798eb22d-9b86-499a-ac8f-24f9c1883594](https://user-images.githubusercontent.com/9307830/129407305-98b51139-9467-4993-a4be-92161875fc3d.png)

Vs other editors with highlighted escaped literals:

![126541203-798eb22d-9b86-499a-ac8f-24f9c1883594](https://user-images.githubusercontent.com/9307830/129407408-b9d26ce4-2354-43d1-bb0e-f0b4e558a4e0.png)

**Comment by tek** [comment]
in general, you can use injections to defer to a different grammar, by putting this into `$VIMCONFIG/queries/haskell/injections.scm`:

```
(string) @grammar-with-escape-highlights
```

And I can look into parsing strings directly in the grammar.

**Comment by tek** [comment]
so, the other grammar would need to be specifically parsing only the string

**Comment by jpe90** [comment]
Got it working after I realized you didn't mean to literally add `@grammar-with-escape-highlights`, LOL. Thank you!!

**Comment by tek** [comment]
oh really, which grammar did you use?

**Comment by jpe90** [comment]
I just tried a couple at random and went with Rust because it picked up \" and \n without giving me problems in the project I'm working on

**Comment by tek** [comment]
oh heh. you could create a minimal grammar that only parses strings tho

**Comment by jpe90** [comment]
I'm not really familiar with treesitter internals (or parsing in general really) to know how to go about that.. would reading https://tree-sitter.github.io/tree-sitter/creating-parsers get me started, or is there other documentation you would also recommend?

**Comment by tek** [comment]
I did a similar thing for quasiquotes here: https://github.com/tek/vim-bundle-conf/tree/master/tree-sitter/exon

it's bare bones, you'll just have to modify `grammar.js`, it should be fairly intuitive, using that page you linked.

then you'll have to add it in you vim config in a `plugin/foo.vim`:

```
lua <<EOF
local parser_config = require "nvim-treesitter.parsers".get_parser_configs()
parser_config.string_grammar_name = {
  install_info = {
    url = "/path/to/string/grammar/",
    files = {"src/parser.c"}
  }
}
EOF
```

and you can refer to it in an injection with `@string_grammar_name`

**Comment by jpe90** [comment]
Thank you, that example was very helpful. I ended up giving it a shot. I set up a minimal grammar which only parses strings into two nodes, "string_fragment" and "escape_sequence". It installs successfully and are applied in the right places when I look at a source file with TSPlayground:

![escape_sequence](https://user-images.githubusercontent.com/9307830/129470970-05a80527-cbd0-4a3a-a767-8966e2aa3680.png)


However I can't seem to get the capture groups set up for them. In the minimal grammar, I have a queries/highlights.scm containing:

```
(string_fragment) @string
(escape_sequence) @escape
```
and a corresponding "escape" custom capture group set up in my vim config.

But those don't seem to get picked up: 

![no_captures_found](https://user-images.githubusercontent.com/9307830/129471110-02783ad6-591a-49be-a220-aa75462b7ee6.png)

I'm sure I'm missing something on how capture groups work, so I'll try to figure it out, but let me know if anything obvious jumps out that I'm missing.

**Comment by tek** [comment]
by "in the minimal grammar", do you mean that the highlights file is in the `porcupine/queries` directory? because that won't get picked up by nvim I think.

**Comment by jpe90** [comment]
Yeah, I had tried adding those highlights in the haskell queries directory in the fork of nvim-treesitter that my config is hooked up to and treesitter exploded with an error referencing an "invalid node". If I included a `porcupine` highlight group it got picked up, though. Anyways I should probably find time to figure out how tree-sitter actually works before trying to hack something together :^) my attempt is [here](https://github.com/jpe90/porcupine)

**Comment by tek** [comment]
well, for nvim, you have to put all the queries and injections into the nvim config. so either your nvim-treesitter fork or your local `~/.config/nvim`.
and the injection has to be in `queries/haskell/injections.scm`, while the highlighting for porcupine has to be in `queries/porcupine/highlights.scm`.

**Comment by jpe90** [comment]
No, I got stuck getting highlights/injections recognized and felt a little out of my depth. I wanted to make more of an effort to understand how they work before I kept bugging for help. I'll give it another shot when I'm feeling brave :^)

**Comment by tek** [comment]
gotcha, feel free to continue here when you need more input!
